making pbr work

bonjorno7 opened the pull request:
oh boy
the lookup table was wrong, I replaced it with an approximate function that actually works
direct lighting wasn't working because of a weird condition in the for loop
implemented ambient occlusion, multiplies over the combined lighting
implemented emission texture, adds to the combined lighting

bonjorno7 committed:
making pbr work

oh boy
the lookup table was wrong, I replaced it with an approximate function that actually works
direct lighting wasn't working because of a weird condition in the for loop
implemented ambient occlusion, multiplies over the combined lighting
implemented emission texture

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_dx9.cpp:
```diff
@@ -250,10 +248,10 @@ END_SHADER_PARAMS
 				BindTexture(SHADER_SAMPLER10, info.mraoTexture, 0);
-			else
-				pShaderAPI->BindStandardTexture(SHADER_SAMPLER10, TEXTURE_WHITE);
+	  else
```
Would you mind using tab indents here as well?

bonjorno7 replied:
done

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps30.fxc:
```diff
@@ -251,41 +277,39 @@ float4 main( PS_INPUT i ) : COLOR
 		directLighting += calculateLight(LightIn, LightColor, outgoingLightDirection,
 			normal, fresnelReflectance, roughness, metalness, lightDirectionAngle, albedo);
+
+		//return float4(directLighting.x + directLighting.y + directLighting.z, 0.0, 0.5, 1.0); // testing if the loop runs at all
```
Do we still need this?

bonjorno7 replied:
nope, removed now

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps30.fxc:
```diff
@@ -251,41 +277,39 @@ float4 main( PS_INPUT i ) : COLOR
 	float3 directLighting = 0.0;
 
-	for (uint n = 0; n < NUM_LIGHTS && n < 0; ++n)
+	for (uint n = 0; n < NUM_LIGHTS; ++n) // now lighting works on one model but not on another??
```
Is this due to the flashlight thing?

bonjorno7 replied:
nope, it's the same with or without the flashlight on, I just can't get direct lighting to work consistently
perhaps I can fix that in the next update
at least it's better than it was :/

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps30.fxc:
```diff
@@ -321,7 +345,9 @@ float4 main( PS_INPUT i ) : COLOR
 
-	float alpha = 1.0;
+	//float alpha = 1.0;
+	float alpha = albedo.a; // Gocnak told me to add this
```
This is ok, can you remove the comment above and to the right?

bonjorno7 replied:
done

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps30.fxc:
```diff
@@ -333,5 +359,11 @@ float4 main( PS_INPUT i ) : COLOR
 
-	return FinalOutput( float4( directLighting + ambientLighting, alpha ), fogFactor, PIXELFOGTYPE, TONEMAP_SCALE_LINEAR, bWriteDepthToAlpha, i.projPos.z );	
+#if EMISSION
+	float3 combinedLighting = (directLighting + ambientLighting) * ambientOcclusion + emission;
```
Would you mind setting `combinedLighting` above the `#if` and doing `combinedLighting += emission;` inside the `#if`?

bonjorno7 replied:
alright

ksharperd commented:
That good,you finally repair it!

bonjorno7 committed:
small changes to pbr

tabs instead of spaces in some places I missed
removed no longer needed testing lines
changed comment to my current understanding of the problem
#if thing at the end is different as requested
made hardcoded max envmap mipmap a static const float cause it's used 3 times

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps30.fxc:
```diff
@@ -25,6 +26,7 @@
 
 #include "common_flashlight_fxc.h"
 #include "common_lightmappedgeneric_fxc.h"
+#include "common_vertexlitgeneric_dx9.h"
```
Why do we need this extra include?

bonjorno7 replied:
we don't, yeet

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps30.fxc:
```diff
@@ -33,16 +35,14 @@ const float3 cAmbientCube[6]				: register( PSREG_AMBIENT_CUBE );
-#define EPSILON							0.00001
+static const float PI = 3.141592;
+static const float EPSILON = 0.00001;
+static const float ENVMAPLOD = 6.0;
```
Could you add a comment explaining what the 6.0 stands for? (1 roughness == 4x4 texture, 0 roughness == 256x256 texture)

bonjorno7 replied:
// 6 is the lowest cubemap mipmap we need. in a 256x256 cubemap, it gives us 4x4

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps30.fxc:
```diff
@@ -154,7 +156,7 @@ float3 calculateLight(float3 lightIn, float3 lightIntensity, float3 lightOut, fl
 
 	// Diffuse scattering happens due to light being refracted multiple times by a dielectric medium.
-	// Metals on the other hand either reflect or absorb energy, so diffuse contribution is always zero.
+	// Metals on the other hand either reflect or absorb energso diffuse contribution is alwaysy,  zero.
```
could you revert or fix these changes?

bonjorno7 replied:
dunno how that happened, fixed

thexa4 commented on mp/src/materialsystem/stdshaders/pbr_ps30.fxc:
```diff
@@ -321,17 +330,19 @@ float4 main( PS_INPUT i ) : COLOR
-	//return tex2D(LightmapSampler, g_LightmapCoords[0].xy + i.baseTexCoord * g_LightmapCoords[0].z);
 
-	return FinalOutput( float4( directLighting + ambientLighting, alpha ), fogFactor, PIXELFOGTYPE, TONEMAP_SCALE_LINEAR, bWriteDepthToAlpha, i.projPos.z );	
+	float3 combinedLighting = (directLighting + ambientLighting) * ambientOcclusion;
```
Ambient occlusion should only apply to `ambientLighting`, not to both.

bonjorno7 replied:
yep

bonjorno7 committed:
more small changes to pbr

first attempt at lightmaps for models (still not working but doesn't crash the game)
added some comments
fixed a typo
renamed LIGHTMAPPED to USE_LIGHTMAPS so it wouldn't be confusing in the C++ file having LIGHTMAPPED refer to two different things
multiply ambientLighting with ambientOcclusion in the ambient term instead of while combining with directLighting

bonjorno7 committed:
Hopefully improved

removed lightmap support on models for now
fixed a bunch of things
bumped lightmaps on brushes work now
every problem that I could find is fixed
except the flashlight bug
maybe next commit :/

bonjorno7 committed:
minor cleanup

slight improvements in readability
double checked everything
fixed sign_det in the pixel shader
